osu.Game/Overlays/Mods/AddPresetButton.cs

```diff
@@ -0,0 +1,164 @@
+            else
+                this.HidePopover();
+        }
+
+        public Popover GetPopover() => new AddPresetPopover(this);
+
+        private class AddPresetPopover : OsuPopover
+        {
+            private readonly AddPresetButton button;
+
+            private readonly LabelledTextBox nameTextBox;
+            private readonly LabelledTextBox descriptionTextBox;
+            private readonly ShearedButton createButton;
+
+            [Resolved]
+            private Bindable<RulesetInfo> ruleset { get; set; } = null!;
+
+            [Resolved]
+            private Bindable<IReadOnlyList<Mod>> selectedMods { get; set; } = null!;
+
+            [Resolved]
+            private RealmAccess realm { get; set; } = null!;
+
+            public AddPresetPopover(AddPresetButton addPresetButton)
+            {
+                button = addPresetButton;
+
+                Child = new FillFlowContainer
+                {
+                    Width = 300,
+                    AutoSizeAxes = Axes.Y,
+                    Spacing = new Vector2(7),
+                    Children = new Drawable[]
+                    {
+                        nameTextBox = new LabelledTextBox
+                        {
+                            Anchor = Anchor.TopCentre,
+                            Origin = Anchor.TopCentre,
+                            Label = CommonStrings.Name,
+                            TabbableContentContainer = this
+                        },
+                        descriptionTextBox = new LabelledTextBox
+                        {
+                            Anchor = Anchor.TopCentre,
+                            Origin = Anchor.TopCentre,
+                            Label = CommonStrings.Description,
+                            TabbableContentContainer = this
+                        },
+                        createButton = new ShearedButton
+                        {
+                            Anchor = Anchor.TopCentre,
+                            Origin = Anchor.TopCentre,
+                            Text = ModSelectOverlayStrings.AddPreset,
+                            Action = tryCreatePreset
+                        }
+                    }
+                };
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(OverlayColourProvider colourProvider, OsuColour colours)
+            {
+                Body.BorderThickness = 3;
+                Body.BorderColour = colours.Orange1;
+
+                createButton.DarkerColour = colours.Orange1;
+                createButton.LighterColour = colours.Orange0;
+                createButton.TextColour = colourProvider.Background6;
+            }
+
+            private void tryCreatePreset()
+            {
+                if (string.IsNullOrWhiteSpace(nameTextBox.Current.Value))
+                {
+                    Body.Shake();
+                    return;
+                }
+
+                realm.Write(r => r.Add(new ModPreset
```

Comment:
Something we can probably leave for the future, but these calls can be `async` to avoid any potential blocking on database operations:

```diff
diff --git a/osu.Game/Overlays/Mods/AddPresetButton.cs b/osu.Game/Overlays/Mods/AddPresetButton.cs
index e3aeb21f1d..8b59093b8f 100644
--- a/osu.Game/Overlays/Mods/AddPresetButton.cs
+++ b/osu.Game/Overlays/Mods/AddPresetButton.cs
@@ -3,6 +3,7 @@
 
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Extensions;
@@ -19,6 +20,7 @@
 using osu.Game.Rulesets.Mods;
 using osuTK;
 using osu.Game.Localisation;
+using osu.Game.Online.Multiplayer;
 
 namespace osu.Game.Overlays.Mods
 {
@@ -117,7 +119,7 @@ public AddPresetPopover(AddPresetButton addPresetButton)
                             Anchor = Anchor.TopCentre,
                             Origin = Anchor.TopCentre,
                             Text = ModSelectOverlayStrings.AddPreset,
-                            Action = tryCreatePreset
+                            Action = () => tryCreatePreset().FireAndForget()
                         }
                     }
                 };
@@ -134,7 +136,7 @@ private void load(OverlayColourProvider colourProvider, OsuColour colours)
                 createButton.TextColour = colourProvider.Background6;
             }
 
-            private void tryCreatePreset()
+            private async Task tryCreatePreset()
             {
                 if (string.IsNullOrWhiteSpace(nameTextBox.Current.Value))
                 {
@@ -142,7 +144,7 @@ private void tryCreatePreset()
                     return;
                 }
 
-                realm.Write(r => r.Add(new ModPreset
+                await realm.WriteAsync(r => r.Add(new ModPreset
                 {
                     Name = nameTextBox.Current.Value,
                     Description = descriptionTextBox.Current.Value,
```

But there would need to be a "loading/processing" state and it's probably not worth it for something like this, until we actually begin to get reports of blocking on database ops. Also the `FireAndForget` method is a bit localised right now.

Reply:
I'm aware of your preference for the async realm variants, but I internally cringe a little every time I have to jam something async inside an `Action`, what with it being very `void`. We could probably do with a button that accepts a task and handles its return properly at some point.

Reply:
Yeah, for sure. Definitely something for a separate effort, and done in bulk for all such user interactions.